### PR TITLE
periph/gpio: merged dir and pushpull parameters

### PR DESCRIPTION
**Pull request by haukepetersen**
This PR solves #4472

Please have a look at the interface change first (commit 57ba62e) -> this is what came out of the discussion in #4472 (or what I interpreted of it :-) ).

~~I adjusted all CPUs except the `kinetis_common`. For the latter I am waiting on #4830 to prevent duplicate work...~~

If you all agree, I guess some extensive testing is in order...

### CODE REVIEW

File: cpu/atmega2560/periph/gpio.c [hunk 1]
```diff
@@ -86,51 +86,36 @@ static inline uint16_t _pin_addr(gpio_t pin)
     return (_port_addr(pin) - 0x02);
 }
 
-int gpio_init(gpio_t pin, gpio_dir_t dir, gpio_pp_t pullup)
+int gpio_init(gpio_t pin, gpio_mode_t mode)
 {
-    int res;
-
-    if (dir == GPIO_DIR_OUT) {
+    if (mode == GPIO_OUT) {
```
Comment (gebart) [inline cpu/atmega2560/periph/gpio.c]: Using `switch(mode){ case xyz:}` instead of `if()` would improve the readability because of the similarities with other periph drivers.

Comment (haukepetersen) [inline cpu/atmega2560/periph/gpio.c]: true. But it is true, that the compiler ends up with the same binary in this case, right?

File: cpu/atmega2560/periph/gpio.c [hunk 2]
```diff
@@ -86,51 +86,36 @@ static inline uint16_t _pin_addr(gpio_t pin)
...
+    else {
+        return -1;
+    }
+    return 0;
 }
 
-int gpio_init_int(gpio_t pin, gpio_pp_t pullup, gpio_flank_t flank,
-                  gpio_cb_t cb, void *arg)
+int gpio_init_int(gpio_t pin, gpio_mode_t mode, gpio_flank_t flank,
+                    gpio_cb_t cb, void *arg)
```
Comment (gebart) [inline cpu/atmega2560/periph/gpio.c]: remove the added whitespaces, alignment is messed up

File: cpu/atmega2560/periph/gpio.c [hunk 3]
```diff
@@ -86,51 +86,36 @@ static inline uint16_t _pin_addr(gpio_t pin)
...
-         || (_port_num(pin) == PORT_E && pin_num < 4)) {
+         || (_port_num(pin) == PORT_E && pin_num < 4)
+         || (mode >= GPIO_OUT)) {
         return -1;
     }
 
-    if (gpio_init(pin, GPIO_DIR_IN, pullup) < 0) {
+    if (gpio_init(pin, mode) < 0) {
         return -1;
     }
```
Comment (gebart) [inline cpu/atmega2560/periph/gpio.c]: The `cli` and `sei` below should be changed to disableIRQ and restoreIRQ(mask), but since that is in master we could make it a separate PR.

File: cpu/atmega2560/periph/gpio.c [hunk 4]
```diff
@@ -86,51 +86,36 @@ static inline uint16_t _pin_addr(gpio_t pin)
...
-                  gpio_cb_t cb, void *arg)
+int gpio_init_int(gpio_t pin, gpio_mode_t mode, gpio_flank_t flank,
+                    gpio_cb_t cb, void *arg)
 {
     uint8_t pin_num = _pin_num(pin);
 
     if ((_port_num(pin) == PORT_D && pin_num > 3)
-         || (_port_num(pin) == PORT_E && pin_num < 4)) {
+         || (_port_num(pin) == PORT_E && pin_num < 4)
+         || (mode >= GPIO_OUT)) {
```
Comment (gebart) [inline cpu/atmega2560/periph/gpio.c]: I don't like this greater than comparison with an enum since it will silently break if the order of the enum values in the global periph/gpio.h file changes. Is there a big penalty on code size if we change this into the below?

```c
switch(mode) {
  case GPIO_IN:
  case GPIO_IN_PU:
  /* (case GPIO_IN_PD:) since atmega2560 does not seem to support pull-down */
  break;
  default:
  return -1;
```

File: cpu/cc2538/periph/gpio.c [hunk 5]
```diff
@@ -360,28 +354,42 @@ int gpio_init(gpio_t dev, gpio_dir_t dir, gpio_pp_t pushpull)
     pin = pin_lut[dev];
     gpio_software_control(pin);
 
-    if (dir == GPIO_DIR_OUT) {
+    if (mode == GPIO_OUT) {
```
Comment (gebart) [inline cpu/cc2538/periph/gpio.c]: `switch(mode)` would improve readability

File: cpu/lpc11u34/periph/gpio.c [hunk 6]
```diff
@@ -279,26 +279,31 @@ int gpio_init(gpio_t dev, gpio_dir_t dir, gpio_pp_t pullup)
 
     /* Disable resistors */
     *lpc_pin_registers[pin + (port * 24)] &= ~(3 << 3);
-    /* Set resistors */
-    if (pullup == GPIO_PULLUP) {
-        *lpc_pin_registers[pin + (port * 24)] |= (2 << 3);
+
+    /* Set mode */
+    if (mode == GPIO_OUT) {
```
Comment (gebart) [inline cpu/lpc11u34/periph/gpio.c]: switch

File: cpu/nrf5x_common/include/periph_cpu.h [hunk 7]
```diff
@@ -54,15 +54,18 @@ extern "C" {
...
  * @{
  */
-#define HAVE_GPIO_PP_T
+#define HAVE_GPIO_MODE_T
 typedef enum {
-    GPIO_NOPULL = 0,        /**< do not use internal pull resistors */
-    GPIO_PULLUP = 3,        /**< enable internal pull-up resistor */
-    GPIO_PULLDOWN = 1       /**< enable internal pull-down resistor */
-} gpio_pp_t;
+    GPIO_IN    = (0 << 0 | 0 << 1 | 0 << 2),    /**< IN */
```
Comment (gebart) [inline cpu/nrf5x_common/include/periph_cpu.h]: What is the RIOT stance on binary literals (`0b00000000`)? It would make sense to use those here.

File: cpu/nrf5x_common/include/periph_cpu.h [hunk 8]
```diff
@@ -54,15 +54,18 @@ extern "C" {
...
+    GPIO_IN_PD = (0 << 0 | 0 << 1 | 1 << 2),    /**< IN with pull-down */
+    GPIO_IN_PU = (0 << 0 | 0 << 1 | 2 << 2),    /**< IN with pull-up */
```
Comment (gebart) [inline cpu/nrf5x_common/include/periph_cpu.h]: the `2 << 2` part is really easy to miss, maybe change this to `(0 << 0 | 0 << 1 | 0 << 2 | 1 << 3)` ?

File: cpu/nrf5x_common/include/periph_cpu.h [hunk 9]
```diff
@@ -54,15 +54,18 @@ extern "C" {
...
+    GPIO_OUT   = (1 << 0 | 1 << 1 | 0 << 2),    /**< OUT (push-pull) */
+    GPIO_OD    = (0xff),                        /**< not supported */
```
Comment (gebart) [inline cpu/nrf5x_common/include/periph_cpu.h]: not supported by hardware or not supported by the current driver implementation?

File: cpu/sam3/include/periph_cpu.h [hunk 10]
```diff
@@ -60,16 +60,25 @@ typedef uint32_t gpio_t;
  */
 #define CPUID_LEN           (16U)
 
+
```
Comment (gebart) [inline cpu/sam3/include/periph_cpu.h]: delete this extra newline

File: cpu/sam3/include/periph_cpu.h [hunk 11]
```diff
@@ -60,16 +60,25 @@ typedef uint32_t gpio_t;
...
 /**
- * @brief Override values for pull register configuration
+ * @brief   Override GPIO modes
+ *
+ * We use 4 bit to determine the pin functions:
```
Comment (gebart) [inline cpu/sam3/include/periph_cpu.h]: 3 bits

File: cpu/sam3/periph/gpio.c [hunk 12]
```diff
@@ -157,31 +157,33 @@ int gpio_init(gpio_t pin, gpio_dir_t dir, gpio_pp_t pushpull)
...
-            break;
-        case GPIO_NOPULL:
-            port->PIO_PUDR = (1 << pin_num);
-            break;
-    }
-    if (dir == GPIO_DIR_OUT) {
-        /* configure pin as output */
+
+    /* configure pin direction (in/out) */
+    if (mode & 0x1) {
```
Comment (gebart) [inline cpu/sam3/periph/gpio.c]: change these literals to preprocessor defines

File: cpu/sam3/include/periph_cpu.h (hunk 11 above)
Comment (gebart) [inline cpu/sam3/include/periph_cpu.h]: Change these to use preprocessor macros or something symbolic as the current implementation contains magic numbers in the code.

File: cpu/samd21/include/periph_cpu.h [hunk 13]
```diff
@@ -35,6 +35,26 @@ enum {
...
+ *
+ * We use 3 bit to determine the pin functions:
+ * - bit 0: PU or PU
+ * - bit 1: input enable
+ * - bit 2: pull enable
+ * @{
+ */
+#define HAVE_GPIO_MODE_T
+typedef enum {
+    GPIO_IN    = 0x2,           /**< IN */
```
Comment (gebart) [inline cpu/samd21/include/periph_cpu.h]: change these to use preprocessor macros or some other kind of symbolic name to avoid magic numbers in the implementation.

File: cpu/saml21/include/periph_cpu.h [hunk 14]
```diff
@@ -33,6 +33,26 @@ enum {
...
+ *
+ * We use 3 bit to determine the pin functions:
+ * - bit 0: PU or PU
+ * - bit 1: input enable
+ * - bit 2: pull enable
+ * @{
+ */
+#define HAVE_GPIO_MODE_T
+typedef enum {
+    GPIO_IN    = 0x2,           /**< IN */
```
Comment (gebart) [inline cpu/saml21/include/periph_cpu.h]: magic numbers

File: cpu/stm32f0/periph/gpio.c [hunk 15]
```diff
@@ -61,31 +61,32 @@ static inline int _pin_num(gpio_t pin)
...
-        port->OTYPER &= ~(1 << pin_num);        /* set to push-pull */
-        port->OSPEEDR |= (3 << (2 * pin_num));  /* set to high speed */
-        port->ODR &= ~(1 << pin_num);           /* set pin to low signal */
-    }
-    else {
-        port->MODER &= ~(3 << (2 * pin_num));   /* configure pin as input */
-    }
+
+    /* set mode */
+    port->MODER &= ~(0x3 << (2 * pin_num));
```
Comment (gebart) [inline cpu/stm32f0/periph/gpio.c]: magic numbers

File: cpu/stm32f1/periph/gpio.c [hunk 16]
```diff
@@ -68,38 +68,32 @@ static inline int _pin_num(gpio_t pin)
...
-        /* clear pin */
-        port->BRR = (1 << pin_num);
-    }
-    else {
-        /* configure pin to input, pull register according to the value of
-         * the pullup parameter */
-        port->CR[pin_num >> 3] |= ((pullup & 0xc)  << ((pin_num & 0x7) * 4));
-        port->ODR &= ~(1 << pin_num);
-        port->ODR |=  ((pullup & 0x1) << pin_num);
+    port->CR[pin_num >> 3] |=  (mode << ((pin_num & 0x7) * 4));
```
Comment (gebart) [inline cpu/stm32f1/periph/gpio.c]: Could you introduce a port number macro/inline to find the index of the CR register array?

File: tests/periph_gpio/main.c [hunk 17]
```diff
@@ -29,101 +29,81 @@ static void cb(void *arg)
...
     if (argc < 4) {
         printf("usage: %s <port> <pin> <flank> [pull_config]\n", argv[0]);
-        puts("      flank:       0: falling\n"
-             "                   1: rising\n"
-             "                   2: both\n"
-             "      pull_config: 0: no pull resistor (default)\n"
-             "                   1: pull up\n"
-             "                   2: pull down");
+        puts("\tflank:\n"
+             "\t0: falling\n"
```
Comment (gebart) [inline tests/periph_gpio/main.c]: `\t\t` instead?

**Comment by gebart** [comment]
ACK for the general interface change. However, I don't like the magic numbers (bitmasks) used in the individual implementations (nrf5x, sam3, saml21, samd21, all stm32). Could you introduce some symbolic names for them so that we avoid future mistakes like https://github.com/RIOT-OS/RIOT/pull/4725#discussion-diff-52355698 ?

File: tests/periph_gpio/main.c (hunk 17 above)
Comment (haukepetersen) [inline tests/periph_gpio/main.c]: no, `\t0` was right, as it should display `        0: falling` in that line

File: cpu/atmega2560/periph/gpio.c [hunk 18]
```diff
@@ -86,51 +86,36 @@ static inline uint16_t _pin_addr(gpio_t pin)
...
 
-int gpio_init(gpio_t pin, gpio_dir_t dir, gpio_pp_t pullup)
+int gpio_init(gpio_t pin, gpio_mode_t mode)
 {
-    int res;
-
-    if (dir == GPIO_DIR_OUT) {
+    if (mode == GPIO_OUT) {
         _SFR_MEM8(_ddr_addr(pin)) |= (1 << _pin_num(pin));
-        res = bit_is_set(_SFR_MEM8(_ddr_addr(pin)), _pin_num(pin));
```
Comment (haukepetersen) [inline cpu/atmega2560/periph/gpio.c]: Nope, there is no reason for this check what so ever, don't ask how this slipped through review...

File: cpu/atmega2560/periph/gpio.c (hunk 3 above)
Comment (haukepetersen) [inline cpu/atmega2560/periph/gpio.c]: agree, unrelated to this PR, separate PR is the way to go.

File: cpu/atmega2560/periph/gpio.c (hunk 4 above)
Comment (haukepetersen) [inline cpu/atmega2560/periph/gpio.c]: I don't quite understand how this is effecting the global `periph/gpio.h` file. The mode type is defined locally by this CPU... 

Using the switch yields 4 byte more code. Allowing only the two 'good' mode specifically (IN and IN_PU) and dropping the check around 'gpio_init()' yields 2 byte less code size -> so that is the way to go then I'd say.

File: cpu/nrf5x_common/include/periph_cpu.h (hunk 7 above)
Comment (haukepetersen) [inline cpu/nrf5x_common/include/periph_cpu.h]: Actually I like the shift notation as chosen quite a bit, though I forgot to write a comment above describing the bits as I did e.g. for the stm32x CPUs

File: cpu/nrf5x_common/include/periph_cpu.h (hunk 8 above)
Comment (haukepetersen) [inline cpu/nrf5x_common/include/periph_cpu.h]: no, as bit 3+4 together describe the pull resistor configuration and the values 0, 1, 2 are the ones also used in the ref manual, I think this notation is quite straight forward!

File: cpu/nrf5x_common/include/periph_cpu.h (hunk 7 above)
Comment (haukepetersen) [inline cpu/nrf5x_common/include/periph_cpu.h]: description added

File: cpu/nrf5x_common/include/periph_cpu.h (hunk 9 above)
Comment (haukepetersen) [inline cpu/nrf5x_common/include/periph_cpu.h]: by hardware

File: cpu/nrf5x_common/periph/gpio.c [hunk 19]
```diff
@@ -38,16 +38,20 @@
 static gpio_isr_ctx_t exti_chan;
 
 
-int gpio_init(gpio_t pin, gpio_dir_t dir, gpio_pp_t pullup)
+int gpio_init(gpio_t pin, gpio_mode_t mode)
 {
+    /* filter out inapplicable modes */
+    if ((mode == GPIO_OD) || (mode == GPIO_OD_PU)) {
```
Comment (haukepetersen) [inline cpu/nrf5x_common/periph/gpio.c]: changed.

File: cpu/nrf5x_common/include/periph_cpu.h (hunk 8 above)
Comment (haukepetersen) [inline cpu/nrf5x_common/include/periph_cpu.h]: but making it more readable, that I can do :-)

File: cpu/nrf5x_common/include/periph_cpu.h (hunk 7 above)
Comment (OlegHahm) [inline cpu/nrf5x_common/include/periph_cpu.h]: Personally, I would prefer hexadecimal or binary representation, but if you prefer that I can live with it as long as you add parens.

Comment (cgundogan) [inline cpu/nrf5x_common/include/periph_cpu.h]: binary representation is not really part of the standard. AFAIK, they are only gcc extensions (clang does also support them). I think using them would work for most compilers out there, but you never know..

**Comment by haukepetersen** [comment]
addressed comments and tried to remove the newly introduced magic numbers. Concerning the existing code and magic numbers: I don't see it as part of this PR to clean that up and I think this should go into a separate one. So I won't touch the stm-gpio implementation for now.

File: cpu/nrf5x_common/include/periph_cpu.h [hunk 20]
```diff
@@ -54,15 +54,33 @@ extern "C" {
...
 /**
- * @brief   Override GPIO pull register select values
+ * @brief   Generate GPIO mode bitfields
+ *
+ * We use 4 bit to encode the pin mode:
+ * - bit   0: output enable
+ * - bit   1: input connect
+ * - bit 2+3: pull resistor configuration
+ */
+#define GPIO_MODE(oe, ic, pr)   (oe << 0 | ic << 1 | pr << 2)
```
Comment (OlegHahm) [inline cpu/nrf5x_common/include/periph_cpu.h]: parens?

File: cpu/sam3/include/periph_cpu.h [hunk 21]
```diff
@@ -61,15 +61,28 @@ typedef uint32_t gpio_t;
...
 /**
- * @brief Override values for pull register configuration
+ * @brief   Generate GPIO mode bitfields
+ *
+ * We use 3 bit to determine the pin functions:
+ * - bit 0: in/out
+ * - bit 1: PU enable
+ * - bit 2: OD enable
+ */
+#define GPIO_MODE(io, pu, od)   (io << 0 | pu << 1 | od << 2)
```
Comment (OlegHahm) [inline cpu/sam3/include/periph_cpu.h]: parens?

**Comment by haukepetersen** [comment]
rebased and added missing parens

File: cpu/stm32l1/include/periph_cpu.h [hunk 22]
```diff
@@ -46,6 +46,31 @@ typedef uint32_t gpio_t;
...
+ * @{
+ */
+#define HAVE_GPIO_MODE_T
+typedef enum {
+    GPIO_IN    = GPIO_MODE(0, 0, 0),    /**< input w/o pull R */
+    GPIO_IN_PD = GPIO_MODE(0, 2, 0),    /**< input with pull-down */
+    GPIO_IN_PU = GPIO_MODE(0, 1, 0),    /**< input with pull-up */
+    GPIO_OUT   = GPIO_MODE(1, 0, 0),    /**< push-pull output */
+    GPIO_OD    = GPIO_MODE(1, 0, 1),    /**< open-drain w/o pull R */
+    GPIO_OD_PU = GPIO_MODE(1, 1, 1)     /**< open-drain with pull-up */
```
Comment (lebrush) [inline cpu/stm32l1/include/periph_cpu.h]: please add `GPIO_OD_PD`

Comment (haukepetersen) [inline cpu/stm32l1/include/periph_cpu.h]: please see my comment in #4472 regarding this.

To clarify: the mean of the periph driver interface is not, to allow for any MCU configuration you can think of/one particular hardware allows, but to condense the useful and wide-spread features that cover >95% of the use-cases.

**Comment by gebart** [comment]
@haukepetersen I agree regarding existing magic numbers, I opened #4880 as a reminder.

**Comment by haukepetersen** [comment]
Thanks!

**Comment by gebart** [comment]
A kind reminder to rebase once #4830 is merged

**Comment by haukepetersen** [comment]
will do! Let's hope Travis is working fast today :-)

**Comment by gebart** [comment]
#4830 was merged

**Comment by haukepetersen** [comment]
I know, just rebasing and adding the stuff for kinetis and lm4f120 :-)

**Comment by haukepetersen** [comment]
No more dependencies to wait on, so I prepared everything now and did a 'mid-term' squash. Now we just need to review and test everything...

**Comment by haukepetersen** [comment]
rebased, so let's test, OK for:
 - [ ] arm7_common
 - [ ] atmega2560
 - [ ] atmega_common
 - [ ] cc2538
 - [ ] cc430
 - [ ] doc.txt
 - [ ] ezr32wg
 - [ ] k60
 - [ ] k64f
 - [ ] kinetis_common
 - [ ] kw2x
 - [ ] lm4f120
 - [ ] lpc11u34
 - [ ] lpc1768
 - [ ] lpc2387
 - [ ] msp430-common
 - [ ] msp430fxyz
 - [ ] nrf51
 - [ ] nrf52
 - [ ] nrf5x_common
 - [ ] sam21_common
 - [ ] sam3
 - [ ] samd21
 - [ ] saml21
 - [ ] stellaris_common
 - [ ] stm32_common
 - [ ] stm32f0
 - [ ] stm32f1
 - [ ] stm32f3
 - [ ] stm32f4
 - [ ] stm32l1

File: cpu/kinetis_common/periph/gpio.c [hunk 23]
```diff
@@ -30,6 +30,16 @@
 #include "periph/gpio.h"
 
 /**
+ * @brief   Get the OCR reg value from the gpio_mode_t value
+ */
+#define MODE_PCR_MASK       (0x23)
```
Comment (gebart) [inline cpu/kinetis_common/periph/gpio.c]: How about a symbolic definition instead?
```c
#define MODE_PCR_MASK       (PORT_PCR_ODE_MASK | PORT_PCR_PE_MASK | PORT_PCR_PS_MASK)
```

File: cpu/kinetis_common/include/periph_cpu.h [hunk 24]
```diff
@@ -51,6 +51,32 @@ typedef uint16_t gpio_t;
...
 /**
+ * @brief   Generate GPIO mode bitfields
+ *
+ * We use the following bits to encode the pin mode:
+ * - bit 0: 0 for pull-down or 1 for pull-up
+ * - bit 1: pull register enable (as configured in bit 0)
+ * - bit 5: OD enable
+ * - bit 7: output or input mode
+ */
+#define GPIO_MODE(pu, pe, od, out)   (pu | (pe << 1) | (od << 5) | (out << 7))
```
Comment (gebart) [inline cpu/kinetis_common/include/periph_cpu.h]: can we replace these numbers?
```c
#define GPIO_MODE(pu, pe, od, out)   (pu << PORT_PCR_PS_SHIFT | (pe << PORT_PCR_PE_SHIFT) | (od << PORT_PCR_ODE_SHIFT) | (out << 7))
```

Comment (haukepetersen) [inline cpu/kinetis_common/include/periph_cpu.h]: I have to say I am not a big fan of these `_PS_SHIFT` macros - using plain nubmers here is un-ambiguous, well documented, and IMHO easier to read than those macros.

**Comment by haukepetersen** [comment]
@PeterKietzmann, @A-Paul, @thomaseichinger, @punchcard60, @lebrush, @cgundogan: are you guys willing to help test this so we can speed things up? Thanks!

**Comment by lebrush** [comment]
Sure, I'll try to find some time by the EoW and test it on stm32f1 and stm32l1 :-)

**Comment by haukepetersen** [comment]
rebased

**Comment by A-Paul** [comment]
Tested with *tests/periph_gpio* **cc2538dk**:

GPIO in/out tested successfull with onboard switches and LED's. But could only use linear addressing. Have to select e.g. `PC1` with `set 0 17` instead of `set 2 1`.

**Comment by lebrush** [comment]
@haukepetersen I'm afraid I won't be able to test it until end of next week... sorry :-(

**Comment by haukepetersen** [comment]
@A-Paul: the linear addressing issue is related to the GPIO driver implementation as it existed before and luckily unrelated to this PR. But giving this driver a complete overhaul should probably be done at some point...

File: cpu/stm32l1/periph/i2c.c [hunk 25]
```diff
@@ -97,9 +97,9 @@ int i2c_init_master(i2c_t dev, i2c_speed_t speed)
     NVIC_EnableIRQ(i2c_config[dev].er_irqn);
 
     /* configure pins */
-    gpio_init(i2c_config[dev].scl, GPIO_DIR_OUT, GPIO_PULLUP);
+    gpio_init(i2c_config[dev].scl, GPIO_OD_PU);
```
Comment (DipSwitch) [inline cpu/stm32l1/periph/i2c.c]: This should be board dependent, some boards already have a pull-up on the I2C lines, which would reduce the pull-up to 5K instead of 10K.

Same goes for line 102.

File: cpu/stm32f1/periph/i2c.c [hunk 26]
```diff
@@ -141,8 +141,8 @@ static void _pin_config(gpio_t scl, gpio_t sda)
 {
     /* toggle pins to reset analog filter -> see datasheet */
     /* set as output */
-    gpio_init(scl, GPIO_DIR_OUT, GPIO_NOPULL);
-    gpio_init(sda, GPIO_DIR_OUT, GPIO_NOPULL);
+    gpio_init(scl, GPIO_OUT);
```
Comment (DipSwitch) [inline cpu/stm32f1/periph/i2c.c]: Shouldn't this both be GPIO_OUT_OD ?

Comment (gebart) [inline cpu/stm32f1/periph/i2c.c]: Yes, I think so too.

File: cpu/stm32l1/periph/i2c.c (hunk 25 above)
Comment (haukepetersen) [inline cpu/stm32l1/periph/i2c.c]: yes, makes absolutely sense. Will fix

File: cpu/stm32f1/periph/i2c.c (hunk 26 above)
Comment (haukepetersen) [inline cpu/stm32f1/periph/i2c.c]: good catch!

Comment (haukepetersen) [inline cpu/stm32f1/periph/i2c.c]: Actually, `GPIO_OUT` is correct. the `OD` mode is set only after toggling the pins in linex 154-155...

**Comment by haukepetersen** [comment]
rebased.

@everyone: any progress in testing this? Would be great to get this done!

File: cpu/stm32f1/include/periph_cpu.h [hunk 27]
```diff
@@ -44,6 +44,7 @@ typedef uint32_t gpio_t;
 #define GPIO_PIN(x, y)      ((GPIOA_BASE + (x << 10)) | y)
 
 /**
+<<<<<<< HEAD
```
Comment (DipSwitch) [inline cpu/stm32f1/include/periph_cpu.h]: Merge conflict :)

Comment (haukepetersen) [inline cpu/stm32f1/include/periph_cpu.h]: autsch :-)

**Comment by haukepetersen** [comment]
addressed merge conflict issue..

**Comment by DipSwitch** [comment]
Found a wrong OTYPER configuration on the STM32 devices, see your pull-requests. The open-drain configuration wouldn't work :)

**Comment by DipSwitch** [comment]
Anyone found time to test this with the following CPU's?

- [ ] cc430
- [ ] ezr32wg
- [ ] lm4f120
- [ ] lpc11u34
- [ ] lpc2387
- [ ] msp430fxyz

**Comment by haukepetersen** [comment]
rebased

**Comment by haukepetersen** [comment]
So, now only the `lm4f120` and `lpc11u34` are to be tested, but I don't have them here.

@emmanuelsearch: could you give the `ek-lm4f120xl` a quick test?

and who has a 'weio' board flying around? @PeterKietzmann, do you have one?

**Comment by haukepetersen** [comment]
rebased once more

File: cpu/stm32f0/periph/gpio.c [hunk 28]
```diff
@@ -61,31 +61,32 @@ static inline int _pin_num(gpio_t pin)
...
+    port->MODER |=  ((mode & 0x3) << (2 * pin_num));
+    /* set pull resistor configuration */
+    port->PUPDR &= ~(0x3 << (2 * pin_num));
+    port->PUPDR |=  (((mode >> 2) & 0x3) << (2 * pin_num));
+    /* set output mode */
+    port->OTYPER &= ~(1 << pin_num);
+    port->OTYPER &=  (((mode >> 4) & 0x1) << (2 * pin_num));
```
Comment (DipSwitch) [inline cpu/stm32f0/periph/gpio.c]: `GPIO->OTYPER |= (((mode >> 4) & 0x1) << pin_num);` See PR: https://github.com/haukepetersen/RIOT/pull/25

File: cpu/stm32f3/periph/gpio.c [hunk 29]
```diff
@@ -64,33 +61,31 @@ static inline int _pin_num(gpio_t pin)
...
+
+    /* set mode */
+    port->MODER &= ~(0x3 << (2 * pin_num));
+    port->MODER |=  ((mode & 0x3) << (2 * pin_num));
+    /* set pull resistor configuration */
+    port->PUPDR &= ~(0x3 << (2 * pin_num));
+    port->PUPDR |=  (((mode >> 2) & 0x3) << (2 * pin_num));
+    /* set output mode */
+    port->OTYPER &= ~(1 << pin_num);
+    port->OTYPER &=  (((mode >> 4) & 0x1) << (2 * pin_num));
```
Comment (DipSwitch) [inline cpu/stm32f3/periph/gpio.c]: `GPIO->OTYPER |= (((mode >> 4) & 0x1) << pin_num);` See PR: https://github.com/haukepetersen/RIOT/pull/25

File: cpu/stm32f4/periph/gpio.c [hunk 30]
```diff
@@ -62,32 +62,31 @@ static inline int _pin_num(gpio_t pin)
...
+
+    /* set mode */
+    port->MODER &= ~(0x3 << (2 * pin_num));
+    port->MODER |=  ((mode & 0x3) << (2 * pin_num));
+    /* set pull resistor configuration */
+    port->PUPDR &= ~(0x3 << (2 * pin_num));
+    port->PUPDR |=  (((mode >> 2) & 0x3) << (2 * pin_num));
+    /* set output mode */
+    port->OTYPER &= ~(1 << pin_num);
+    port->OTYPER &=  (((mode >> 4) & 0x1) << (2 * pin_num));
```
Comment (DipSwitch) [inline cpu/stm32f4/periph/gpio.c]: `GPIO->OTYPER |= (((mode >> 4) & 0x1) << pin_num);` See PR: https://github.com/haukepetersen/RIOT/pull/25

File: cpu/stm32l1/periph/gpio.c [hunk 31]
```diff
@@ -63,32 +63,31 @@ static inline int _pin_num(gpio_t pin)
...
+
+    /* set mode */
+    port->MODER &= ~(0x3 << (2 * pin_num));
+    port->MODER |=  ((mode & 0x3) << (2 * pin_num));
+    /* set pull resistor configuration */
+    port->PUPDR &= ~(0x3 << (2 * pin_num));
+    port->PUPDR |=  (((mode >> 2) & 0x3) << (2 * pin_num));
+    /* set output mode */
+    port->OTYPER &= ~(1 << pin_num);
+    port->OTYPER &=  (((mode >> 4) & 0x1) << (2 * pin_num));
```
Comment (DipSwitch) [inline cpu/stm32l1/periph/gpio.c]: `GPIO->OTYPER |= (((mode >> 4) & 0x1) << pin_num);` See PR: https://github.com/haukepetersen/RIOT/pull/25

**Comment by haukepetersen** [comment]
fixed otyper for stm32fx cpus

**Comment by haukepetersen** [comment]
also tested successfully for the `ek-lm4f120xl` board. As no-one with access to the `weio` seems to test, I will squash and would be willing to merge without testing for the `lpc11u34`. Also the changes to the `lpc11u34` are quite straight forward and should not break anything.

@DipSwitch, @lebrush, @gebart do you (re-)ACK?

**Comment by haukepetersen** [comment]
rebased and squashed

**Comment by haukepetersen** [comment]
fixed limifrog

**Comment by A-Paul** [comment]
@haukepetersen

> and who has a 'weio' board flying around? @PeterKietzmann, do you have one?

FYI, I asked Peter. We don't have one here.

File: cpu/stm32f1/include/periph_cpu.h [hunk 32]
```diff
@@ -59,6 +59,33 @@ typedef uint32_t gpio_t;
...
+ * @brief   Override GPIO mode options
+ *
+ * We use 4 bit to encode CNF and MODE.
+ * @{
+ */
+#define HAVE_GPIO_MODE_T
+typedef enum {
+    GPIO_IN    = GPIO_MODE(0, 1),   /**< input w/o pull R */
+    GPIO_IN_PD = GPIO_MODE(0, 2),   /**< input with pull-down */
+    GPIO_IN_PU = GPIO_MODE(0, 2),   /**< input with pull-up */
```
Comment (DipSwitch) [inline cpu/stm32f1/include/periph_cpu.h]: It seems improbable that IN_PU and IN_PD should have the same value :)

File: drivers/dht/dht.c [hunk 33]
```diff
@@ -120,7 +120,7 @@ int dht_read(dht_t *dev, int16_t *temp, int16_t *hum)
 
     /* set pin high again - so we can trigger the next reading by pulling it low
      * again */
-    gpio_init(dev->pin, GPIO_DIR_OUT, dev->pull);
+    gpio_init(dev->pin, GPIO_OUT);
```
Comment (DipSwitch) [inline drivers/dht/dht.c]: It is correct that the pullup is not enabled here?

**Comment by DipSwitch** [comment]
Pffeewwww long read 40 minutes or so haha, but seems ok apart from the F1 `gpio_mode_t` definition. :)

**Comment by DipSwitch** [comment]
If the F1 is fixed ACK from my side :)

File: cpu/stm32f1/include/periph_cpu.h (hunk 32 above)
Comment (haukepetersen) [inline cpu/stm32f1/include/periph_cpu.h]: nope, this is correct -> this only enables the pull resistor in general, see `stm32f1/periph/gpio.c` line 89/90, the output register value determines if PU or PD are used.

**Comment by haukepetersen** [comment]
The F1 is actually correct, Travis is happy -> go go go

**Comment by DipSwitch** [comment]
No because now the pull-up will always be enabled ;)

**Comment by haukepetersen** [comment]
No, when you use `GPIO_IN`, the pull resistors are disabled.

File: cpu/stm32f1/periph/gpio.c [hunk 34]
```diff
@@ -68,38 +68,32 @@ static inline int _pin_num(gpio_t pin)
...
+    /* set initial state of output register */
+    port->BRR = (1 << pin_num);
+    if (mode == GPIO_IN_PU) {
```
Comment (DipSwitch) [inline cpu/stm32f1/periph/gpio.c]: because

     +    GPIO_IN_PD = GPIO_MODE(0, 2),   /**< input with pull-down */
     +    GPIO_IN_PU = GPIO_MODE(0, 2),   /**< input with pull-up */

is the same value, this if statement is also true for PD :)

**Comment by DipSwitch** [comment]
Yeah that's true, sorry for the mis-communication. I meant the the pull-up is always enabled, even for pull-down :)

**Comment by haukepetersen** [comment]
Oh, seems I wearing my pink glasses today, you are perfectly right! Will provide a fix ASAP

**Comment by DipSwitch** [comment]
haha np, happens on Friday ;)

**Comment by haukepetersen** [comment]
-> #5104
